### PR TITLE
Add unix socket for daemon access for privileged containers 

### PR DESCRIPTION
**Pull request by crosbymichael**
This adds a separate socket for providing introspection to containers.
It is automatically added to all privileged containers as well as the
current copy of the docker client binary for use with the socket.

The socket will be placed in `/run/docker.sock`.  All linux
systems/images should have a symlink from `/var/run` to `/run`.

The introspection socket does not include the swarm/1.12 endpoints as
these require advanced authentication for use.  The consumers will get a
404 if they try to access a swarm endpoint.

Signed-off-by: Michael Crosby crosbymichael@gmail.com

### CODE REVIEW

**Comment by icecrime** [comment]
I like it :+1: cc @justincormack

**Comment by aluzzardi** [comment]
Overall LGTM

Wondering if this should be `--privileged` or another flag. Although I guess they're interchangeable: if you have access to docker you can easily become privileged, and if you're privileged you can get access to the docker socket anyway.

/cc @diogomonica

**Comment by crosbymichael** [comment]
@aluzzardi ya, it should be another flag as well, waiting for @justincormack to remind me of the good flag name that he thought of, but for testing right now we can add this with privileged containers as it fits the number 1 use case.

**Comment by thaJeztah** [comment]
earlier discussion https://github.com/docker/docker/pull/9135 and https://github.com/docker/docker/pull/9135#issuecomment-67427605 😇

**Comment by diogomonica** [comment]
@crosbymichael what changed? :)

File: container/container_unix.go [hunk 1]
```diff
@@ -407,6 +408,29 @@ func (container *Container) TmpfsMounts() []Mount {
...
+		return nil, nil
+	}
+	path, err := os.Readlink(reexec.Self())
+	if err != nil {
+		return nil, err
+	}
+	return []Mount{
+		{
+			Source:      filepath.Join(filepath.Dir(path), "docker"),
+			Destination: "/usr/bin/docker",
```
Comment (tonistiigi) [inline container/container_unix.go]: Why mount the binary? This is very unlikely to work inside a container with dynamic binaries. Doesn't even work with debian.

File: container/container_unix.go [hunk 2]
```diff
@@ -407,6 +408,29 @@ func (container *Container) TmpfsMounts() []Mount {
...
+			Writable:    true,
+		},
+		{
+			Source:      "/run/docker-introspection.sock",
+			Destination: "/run/docker.sock",
```
Comment (tonistiigi) [inline container/container_unix.go]: Default socket for docker is `/var/run/docker.sock` so this would not work if image doesn't have a `/var/run -> /run` symlink(like busybox).

File: cmd/dockerd/daemon.go [hunk 3]
```diff
@@ -287,24 +239,39 @@ func (cli *DaemonCli) start() (err error) {
 		"graphdriver": d.GraphDriverName(),
 	}).Info("Docker daemon")
 
-	cli.initMiddlewares(api, serverConfig)
-	initRouter(api, d, c)
+	// setup the default routes
+	defaultRoutes := getRouters(d, c)
+	api.InitRouter(utils.IsDebugEnabled(), defaultRoutes...)
+
+	// setup the introspections routes without cluster capabilities
```
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: I don't really like making an exception here. Only reason I can see for not allowing swarm methods here is that user can abuse is or it is not very useful as user can't control if they are on a manager or not. Both or these don't seem valid as other methods are very dangerous as well and this isn't suitable for many other use-cases as well atm because user needs to fully trust the container. We should add more complicated ways of limiting this api in the future(as discussed) but if this remains as fully privileged way to access daemon in global context I don't see why we should forbid `swarm init/join/leave`.

File: container/container_unix.go (hunk 1 above)
Comment (justincormack) [inline container/container_unix.go]: Yes, mounting the binary will not work in general, should not do this (the user can bind mount if they think it will).

File: container/container_unix.go [hunk 4]
```diff
@@ -407,6 +408,29 @@ func (container *Container) TmpfsMounts() []Mount {
...
+			Writable:    true,
```
Comment (justincormack) [inline container/container_unix.go]: I don't think this needs to be writable, you can use a socket rw on a ro mount.

**Comment by thaJeztah** [comment]
Overall, I'm a bit confused; is this PR for *introspection* or for *controlling* docker? For *introspection*, I'd expect a container to be able to obtain information about *itself* (published ports on the host, labels, ...?)

For *controlling* docker, this makes sense, but various discussions in the past lead to a conclusion that we should not do this without fine-grained control over *what* access the container gets. For example, there's many use-cases of containers that only need access to docker events, without the need to be able to *control* docker. Currently those bind-mount the docker socket (which is obviously insecure), some things from the top of my head;

- being able to listen to a (filtered) set of docker events (specific type of events, or only events for specific containers / services), and (related) being able to *inspect* the source of that event
- getting access to docker stats (filtered?)

If we decide on providing full access by default, changing that to a *limited* set may be complicated in future. I'd far more like a limited introspection API to start with, and opening up more features when needed.

I really don't like connecting this to `--privileged`, I think `--privileged` in general was a bad decision (I think @jpetazzo even mentioned he wanted to call it `--insecure` at first); we discourage people to use `--privileged` because it's adding way too much privileges; adding more features to it only seems to *encourage* people to use it (because it's convenient).

Just my 0.02c.

TL;DR; I really like having an introspection API as a feature, but in this form, I don't see it as an improvement over what we currently have

File: container/container_unix.go (hunk 1 above)
Comment (crosbymichael) [inline container/container_unix.go]: isn't it static?

File: cmd/dockerd/daemon.go (hunk 3 above)
Comment (crosbymichael) [inline cmd/dockerd/daemon.go]: for many reasons.  swarm needs more attention or else we create a backdoor and undo many of the security measures that were built in.  It is also inconsistent if you are run on a manager/worker.

**Comment by cpuguy83** [comment]
Won't this break anyone starting docker daemon in a container?

File: cmd/dockerd/daemon.go (hunk 3 above)
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: It's always a backdoor. You can even mount the original docker socket using the "instrospection" socket.

**Comment by crosbymichael** [comment]
@justincormack what was the flag name that you were thinking about for this feature ?

**Comment by justincormack** [comment]
@crosbymichael I forget, its annoying me...

The janky failure looks real, read only rootfs.

@cpuguy83 comment is an issue I think - either we should put the introspection socket on a different path (ugh) or make it optional.

**Comment by crosbymichael** [comment]
when we figureout a good flag name i'll just remove this from `--privileged`

**Comment by cpuguy83** [comment]
For the flag name, given it's just a socket to the daemon and not specifically introspection per se, why not call it as such?
Something like `--daemon-socket /path/to/socket` (or something along those lines).

**Comment by tonistiigi** [comment]
I'd prefer a flag that lets us extend this feature in the future:

```
--cap-add DOCKER_PRIVILEGED_DAEMON
--docker-cap-add PRIVILEGED_DAEMON
--security-opt docker:privileged-daemon
```

**Comment by crosbymichael** [comment]
@ibuildthecloud have you seen this?  Do you have any comments/concerns?

**Comment by crosbymichael** [comment]
Lets close this until we have a better idea on what people expect and want
